src/plumbing.rs

```diff
@@ -0,0 +1,118 @@
+use crate::Database;
+use crate::Query;
+use crate::QueryTable;
+use std::fmt::Debug;
+use std::hash::Hash;
+
+pub use crate::derived::DependencyStorage;
+pub use crate::derived::MemoizedStorage;
+pub use crate::derived::VolatileStorage;
+pub use crate::input::InputStorage;
+pub use crate::runtime::Revision;
+
+pub struct CycleDetected;
```

Comment:
In theory, `CycleDetected` should be public API, but we don't have public cycle-detecting API yet, so it's safer to move it to plumbing for the time being.